src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Listeners/ServiceBusListener.cs

```diff
@@ -177,5 +209,338 @@ private void ThrowIfDisposed()
+            // If detected, this results in an automatic scale out for the site container.
+            if (metrics[0].TotalNewEventCount > 0)
+            {
+                bool queueLengthIncreasing =
+                IsTrueForLastN(
+                    metrics,
+                    NumberOfSamplesToConsider,
+                    (prev, next) => prev.TotalNewEventCount < next.TotalNewEventCount) && metrics[0].TotalNewEventCount > 0;
+                if (queueLengthIncreasing)
+                {
+                    status.Vote = ScaleVote.ScaleOut;
+                    _logger.LogInformation($"Queue length is increasing for '{_entityPath}'.");
+                    return status;
+                }
+            }
+
+            if (metrics[0].EventQueueTime > TimeSpan.Zero && metrics[0].EventQueueTime < metrics[NumberOfSamplesToConsider - 1].EventQueueTime)
+            {
+                bool queueTimeIncreasing =
+                    IsTrueForLastN(
+                        metrics,
+                        NumberOfSamplesToConsider,
+                        (prev, next) => prev.EventQueueTime <= next.EventQueueTime);
+                if (queueTimeIncreasing)
+                {
+                    status.Vote = ScaleVote.ScaleOut;
+                    _logger.LogInformation($"Queue time is increasing for '{_entityPath}'.");
+                    return status;
+                }
+            }
+
+            bool queueLengthDecreasing =
+                IsTrueForLastN(
+                    metrics,
+                    NumberOfSamplesToConsider,
+                    (prev, next) => prev.TotalNewEventCount > next.TotalNewEventCount);
+            if (queueLengthDecreasing)
+            {
+                status.Vote = ScaleVote.ScaleIn;
+                _logger.LogInformation($"Queue length is decreasing for '{_entityPath}'.");
+                return status;
+            }
+
+            bool queueTimeDecreasing = IsTrueForLastN(
+                metrics,
+                NumberOfSamplesToConsider,
+                (prev, next) => prev.EventQueueTime > next.EventQueueTime);
+            if (queueTimeDecreasing)
+            {
+                status.Vote = ScaleVote.ScaleIn;
+                _logger.LogInformation($"Queue time is decreasing for '{_entityPath}'.");
+                return status;
+            }
+
+            _logger.LogInformation($"Service Bus entity '{_entityPath}' is steady.");
+
+            return status;
+        }
+
+        private bool TimeToLogWarning()
+        {
+            DateTime currentTime = DateTime.UtcNow;
+            bool timeToLog = currentTime >= _nextWarningTime;
+            if (timeToLog)
+            {
+                _nextWarningTime = currentTime.AddHours(1);
+            }
+            return timeToLog;
+        }
+
+        private void ParseTopicAndSubscription(string entityPath, out string topic, out string subscription)

```

Comment:
This is the thing I like the least about this implementation. I have to have topic and subscription separated to get the subscription runtime. I wasn't able to find a way to do this in the SDK (only ways to combine them). I did find [this](https://github.com/Azure/azure-service-bus-dotnet/issues/67) github thread that makes it seem like doing it this way would work, but it would break if it ever changed. The only time we have this info separated and available is in `ServiceBusTriggerAttributeBindingProvider` but I thought it would be weird to keep passing that down through the tree, especially since the repo is trying to make these classes generic to work for topics or queues

Reply:
Yeah, you're right. The listener was generalized such that it no longer had queue/topic specific operations in it, hidden behind MessagingProvider.CreateClientEntity for the most part. And CreateClientEntity has similar parsing code in it already